File: 289.GameOfLife/GameOfLife.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        if not board or not board[0]:
            return
        
        rows, cols = len(board), len(board[0])
        
        # Directions for the 8 neighbors
        directions = [(-1, -1), (-1, 0), (-1, 1), 
                      (0, -1),          (0, 1), 
                      (1, -1), (1, 0), (1, 1)]
        
        # Create a copy of the original board
        original = [[board[r][c] for c in range(cols)] for r in range(rows)]
        
        for r in range(rows):
            for c in range(cols):
                live_neighbors = sum(original[r + dr][c + dc] == 1 
                                     for dr, dc in directions 
                                     if 0 <= r + dr < rows and 0 <= c + dc < cols)
                
                if original[r][c] == 1:
                    # Cell is currently alive
                    if live_neighbors < 2 or live_neighbors > 3:
                        board[r][c] = 0
                else:
                    # Cell is currently dead
                    if live_neighbors == 3:
                        board[r][c] = 1 

File: 289.GameOfLife/GameOfLife.ts
/**
 * Do not return anything, modify board in-place instead.
 * @param {number[][]} board
 * @return {void}
 */

function gameOfLife(board) {
    const rows = board.length;
    const cols = board[0].length;
    const directions = [
        [-1, -1], [-1, 0], [-1, 1],
        [0, -1],          [0, 1],
        [1, -1], [1, 0], [1, 1]
    ];

    const countLiveNeighbors = (r, c) => {
        let count = 0;
        for (const [dr, dc] of directions) {
            const nr = r + dr;
            const nc = c + dc;
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && Math.abs(board[nr][nc]) === 1) {
                count++;
            }
        }
        return count;
    };

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            const liveNeighbors = countLiveNeighbors(r, c);
            if (board[r][c] === 1) {
                if (liveNeighbors < 2 || liveNeighbors > 3) {
                    board[r][c] = -1; // Mark as dead
                }
            } else {
                if (liveNeighbors === 3) {
                    board[r][c] = 2; // Mark as alive
                }
            }
        }
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (board[r][c] === -1) {
                board[r][c] = 0; // Convert to dead cell
            } else if (board[r][c] === 2) {
                board[r][c] = 1; // Convert to alive cell
            }
        }
    }
}